### ISSUE TITLE
Clone from private repository

### ISSUE THREAD

**Issue by chamilad** [jelmer/dulwich#249]
I couldn't find any resources on how to clone from a private repository by providing the username and the password. The following code gives an error. It would be a great help if you can add a section on the tutorial on how to do this. 

```
repo = Repo.init("local_repo_path", mkdir=True)
client, remote_path = get_transport_and_path("repo_url", auth=True, username="username", password="password")
```

Error

```
Traceback (most recent call last):
  File "dulwichtest.py", line 54, in <module>
    auth_clone_from_repo("https://chamilad@bitbucket.org/chamilad/testpr.git", "authlocal", "chamilad", "XegKhwMGMlmX")
  File "dulwichtest.py", line 45, in auth_clone_from_repo
    client, remote_path = get_transport_and_path(repo_url, auth=True, username=username, password=password)
  File "/home/chamilad/virtualenv/cartridgeagent/local/lib/python2.7/site-packages/dulwich/client.py", line 1110, in get_transport_and_path
    return get_transport_and_path_from_url(location, **kwargs)
  File "/home/chamilad/virtualenv/cartridgeagent/local/lib/python2.7/site-packages/dulwich/client.py", line 1091, in get_transport_and_path_from_url
    **kwargs), parsed.path
  File "/home/chamilad/virtualenv/cartridgeagent/local/lib/python2.7/site-packages/dulwich/client.py", line 935, in __init__
    GitClient.__init__(self, *args, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'username'
```

Thanks!

**Comment by jelmer** [comment]
On Thu, Jan 22, 2015 at 09:53:37AM -0800, Chamila wrote:
> I couldn't find any resources on how to clone from a private repository by providing the username and the password. The following code gives an error. It would be a great help if you can add a section on the tutorial on how to do this. 

URLs allow specification of usernames and passwords, e.g.:

http://username:password@host/path/to/repo

Cheers,

Jelmer

**Comment by chamilad** [comment]
Hi Jelmer,

Thanks for the prompt response. However specifying the credentials in the URL results in the following error. The `:` is parsed as the port specifier.

```
repo_url = "https://username:password@bitbucket.org/chamilad/testpr.git"
repo = Repo.init(local_repo_path, mkdir=True)
client, remote_path = get_transport_and_path(repo_url)
remote_refs = client.fetch(remote_path, repo)
```

```
Traceback (most recent call last):
  
  File "dulwichtest.py", line 47, in auth_clone_from_repo
    remote_refs = client.fetch(remote_path, repo)
  File "/home/chamilad/virtualenv/cartridgeagent/local/lib/python2.7/site-packages/dulwich/client.py", line 218, in fetch
    progress)
  File "/home/chamilad/virtualenv/cartridgeagent/local/lib/python2.7/site-packages/dulwich/client.py", line 1044, in fetch_pack
    "git-upload-pack", url)
  File "/home/chamilad/virtualenv/cartridgeagent/local/lib/python2.7/site-packages/dulwich/client.py", line 958, in _discover_references
    resp = self._http_request(url, headers)
  File "/home/chamilad/virtualenv/cartridgeagent/local/lib/python2.7/site-packages/dulwich/client.py", line 943, in _http_request
    resp = self.opener.open(req)
  File "/usr/lib/python2.7/urllib2.py", line 404, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 422, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 382, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1222, in https_open
    return self.do_open(httplib.HTTPSConnection, req)
  File "/usr/lib/python2.7/urllib2.py", line 1153, in do_open
    h = http_class(host, timeout=req.timeout) # will parse host:port
  File "/usr/lib/python2.7/httplib.py", line 1164, in __init__
    source_address)
  File "/usr/lib/python2.7/httplib.py", line 704, in __init__
    self._set_hostport(host, port)
  File "/usr/lib/python2.7/httplib.py", line 732, in _set_hostport
    raise InvalidURL("nonnumeric port: '%s'" % host[i+1:])
httplib.InvalidURL: nonnumeric port: 'password@bitbucket.org'

```

Thanks!

**Comment by jelmer** [comment]
Hmm, so actually it seems like we don't support that for HTTP yet. For SSH we also support passing in a username and password independently. Re-opening :)

**Comment by samstav** [comment]
Looks like this might be considered a bug in urllib2.

http://bugs.python.org/issue944396

**Comment by chamilad** [comment]
Hi Jelmer, 

Any updates on when would this be fixed? :) This seems like a crucial setback to be used with private Git repositories. 

Thanks!

**Comment by jsbain** [comment]
Just use a custom urllib2 opener.   Here is what I'm using on another project  (you can probably eliminate the `config` business, frankly I forget why I have that in there)


```python
def auth_urllib2_opener(config, top_level_url, username, password):
    if config is not None:
        proxy_server = config.get("http", "proxy")
    else:
        proxy_server = None

    # create a password manager
        password_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm()

        # Add the username and password.
        # If we knew the realm, we could use it instead of None.
        #top_level_url = "http://example.com/foo/"
        password_mgr.add_password(None, top_level_url, username, password)

        handler = urllib2.HTTPBasicAuthHandler(password_mgr)

    handlers = [handler]
    if proxy_server is not None:
        handlers.append(urllib2.ProxyHandler({"http": proxy_server}))
    opener = urllib2.build_opener(*handlers)
    if config is not None:
        user_agent = config.get("http", "useragent")
    else:
        user_agent = None
    if user_agent is None:
        user_agent = default_user_agent_string()
    opener.addheaders = [('User-agent', user_agent)]
    return opener


client, path = get_transport_and_path(remote_location, **{'config': config} if config else {})

    if type(client) is HttpGitClient and opener:
        client.opener = opener

    remote_refs = client.fetch(path, r, progress=errstream.write)
```

**Comment by chamilad** [comment]
Thanks @jsbain! I was able to tweak your code to achieve cloning from a private repository. 

@jelmer Keeping this open to track solution.

**Comment by jelmer** [comment]
This has been resolved in 0.15.0.
